File: src/java/magma/collect/Joiner.java
package magma.collect;

import magma.option.None;
import magma.option.Option;
import magma.option.Some;

public record Joiner(String delimiter) implements Collector<String, Option<String>> {
    @Override
    public Option<String> createInitial() {
        return new None<>();
    }

    @Override
    public Option<String> fold(Option<String> maybeCurrent, String element) {
        return new Some<>(maybeCurrent.map(inner -> inner + delimiter + element).orElse(element));
    }
}

File: src/clang/magma/collect/stream/Joiner.c
#include "Joiner.h"
int __lambda0__(){return inner;
}
magma.collect.stream.public Joiner(){this("");
}
magma.option.Option<String> createInitial(){return ();
}
magma.option.Option<String> fold(magma.option.Option<String> maybeCurrent, String element){return (maybeCurrent.map(__lambda0__+delimiter+element).orElse(element));
}
